Getting started guide (#162)
bdaiinstitute/synchros2: synchros2/docs/GettingStartedGuide.md

    cd <workspace>
    colcon build --symlink-install
    source install/setup.bash
    ```
    

# Overview

`synchros2` is a framework that allows blocking in callbacks without the user needing to worry about the threading.  The simplest way to use it is to:

1. Decorate your `main` with the `@ros_process.main()` decorator:
    
    ```python
    import synchros2.process as ros_process
    
    @ros_process.main()
    def main() -> None:
    	...
    ```
    
2. Only use the node associated with `ros_scope`:
    
    ```python
    import synchros2.scope as ros_scope
    
    class MyClass:
    	def __init__(self):
    		self._node = ros_scope.ensure_node()
    ```
    

# Your First `synchros2` Code

We’ll now create and run example `synchros2` code.  We’ll analyze the code in the next section.

1. Create a package called `synchros2_tutorials` that depends on `synchros2` (by specifying `--dependencies synchros2` the package creation will automatically add a dependency on `synchros2` to the created `package.xml`): 
    
    ```bash
    cd <workspace>/src
    ros2 pkg create --build-type ament_python --license Apache-2.0 synchros2_tutorials  --dependencies synchros2
    ```
    

> 2. Create the file `<workspace>/src/synchros2_tutorials/synchros2_tutorials/listener.py` and add the following code to it.

Reviewer: @jbarry-bdai nit: 
```suggestion
2. Create the file `<workspace>/src/synchros2_tutorials/synchros2_tutorials/listener.py` and add the following code to it:
```

    
    ```python
    import std_msgs.msg
    import time
    
    import synchros2.process as ros_process
    import synchros2.scope as ros_scope
    
    class Listener:
        def __init__(self):
            self._node = ros_scope.ensure_node()
            self._sub1 = self._node.create_subscription(std_msgs.msg.String, "chat", self._callback1, 1)
            self._sub2 = self._node.create_subscription(std_msgs.msg.String, "chat", self._callback2, 1)
            self._node.get_logger().info("Listening!")
    
        def _callback1(self, msg: std_msgs.msg.String) -> None:
            self._node.get_logger().info(f"Callback 1 received message {msg} and will now sleep for 10 seconds")
            time.sleep(10)
            self._node.get_logger().info(f"Callback 1 is done sleeping after receiving {msg}")
    
        def _callback2(self, msg: std_msgs.msg.String) -> None:
            self._node.get_logger().info(f"Callback 2 received message {msg} and will now sleep for 5 seconds")
            time.sleep(5)
            self._node.get_logger().info(f"Callback 2 is done sleeping after receiving {msg}")
    
    @ros_process.main()
    def main() -> None:
        _ = Listener()

>         ros_process.wait_for_shutdown()

Reviewer: @jbarry-bdai meta: looking at this laid out like that, I feel like we need a better pattern than this construct-and-forget thing (though it is what it is and we would mostly be hiding it). Maybe something like:

```
@ros_process.main(cli())
def main(args):
     ros_process.exec(Listener(args))
```
and then have Listener optionally define a `__call__` override. If it does, it'll get called, if it doesn't, exec will just wait until shutdown.

Reviewer 2: I agree with jenny that its tidier but a lot more confusing for a new user

Reviewer: Fair. I'm not very fond of the constructor-does-things pattern, but I guess folks can implement their own variations of `Listener.run()` if they feel like it. Maybe that's an interesting addition: a tutorial where the process does things at a fixed rate in the main thread (ROS things or not).

    [INFO] [1759264529.884469744] [listener]: Callback 1 is done sleeping after receiving std_msgs.msg.String(data='This is the first message')
    [INFO] [1759264529.885723251] [listener]: Callback 1 received message std_msgs.msg.String(data='This is the second message') and will now sleep for 10 seconds
    [INFO] [1759264530.003322694] [listener]: Callback 2 is done sleeping after receiving std_msgs.msg.String(data='This is the second message')
    [INFO] [1759264539.896067303] [listener]: Callback 1 is done sleeping after receiving std_msgs.msg.String(data='This is the second message')
    
    ```
    

# Analyzing the Code

Let’s first look at what the `listener.py` code is intended to do and then we’ll show why it outputs what it does.

We start with the imports all `synchros2` code needs (non-main code doesn’t need the `ros_process` one):

```python
import synchros2.process as ros_process
import synchros2.scope as ros_scope
```

We then create a class that stores the ros scope node:

```python
class Listener:
    def __init__(self):
        self._node = ros_scope.ensure_node()
```

> The `synchros` mechanisms create a single node that we want to use everywhere because it's already set up to create all the appropriate threads for any callbacks added to it.  We refer to this single node as the "ros scope node".  For beginning use, the ros scope node is the *only* node you need or should use.  The `synchros2` mechanics take care of spinning any callbacks that we attach to this node and spawning whatever threads we need.  This node can be obtained anywhere in the process with the `ros_scope.ensure_node()` call (which will either return the ros scope node or raise a `ValueError` if one was never created - usually because you forgot to decorate your `main` with `ros2_process.main()`).  **Do not create further nodes.**

Reviewer: @jbarry-bdai nit: 
```suggestion
The `synchros2` mechanisms create a single node that we want to use everywhere because it's already set up to create all the appropriate threads for any callbacks added to it.  We refer to this single node as the "ros scope node".  For beginning use, the ros scope node is the *only* node you need or should use.  `synchros2` takes care of spinning any callbacks that we attach to this node and spawning whatever threads we need.  This node can be obtained anywhere in the process with the `ros_scope.ensure_node()` call (which will either return the ros scope node or raise a `ValueError` if one was never created - usually because you forgot to decorate your `main` with `ros2_process.main()`).  **Do not create further nodes.**
```

```python
    def _callback1(self, msg: std_msgs.msg.String) -> None:
        self._node.get_logger().info(f"Callback 1 received message {msg} and will now sleep for 10 seconds")
        time.sleep(10)
        self._node.get_logger().info(f"Callback 1 is done sleeping after receiving {msg}")

    def _callback2(self, msg: std_msgs.msg.String) -> None:
        self._node.get_logger().info(f"Callback 2 received message {msg} and will now sleep for 5 seconds")
        time.sleep(5)
        self._node.get_logger().info(f"Callback 2 is done sleeping after receiving {msg}")

```

Our main function is decorated with `ros_process.main()`:

```python
@ros_process.main()
def main() -> None:
```

**This decorator is very important!**  This is what creates the ros process node and the corresponding executor that keeps track of all of the threads.

Inside the main we create an instance of the listener class and then we wait for shutdown.  This leaves our subscriptions running until the program receives `Ctrl+C`:

```python

>     _ = Listener()x

Reviewer: @jbarry-bdai extra x:

```suggestion
    _ = Listener()
```

> **This decorator is very important!**  This is what creates the ros process node and the corresponding executor that keeps track of all of the threads.

Reviewer: @jbarry-bdai consider: 
```suggestion
**This decorator is very important!**  This is what creates the process-wide ros scope node and the corresponding executor that keeps track of all of the threads.
```
or similar to connect the notions of process node with scope node without explaining much.

2. Immediately after 1: That message went into the queues for both callbacks
3. Immediately after 2: Callback 1 received the first message, printed out the info and started to sleep
4. Immediately after 2: *While callback 1 is running*, callback 2 also received the first message, printed out the info and started to sleep
5. Slightly later: We published the message “This is the second message” on `/chat`
6. Immediately after 5: That message went into the queues for both callbacks.  However the callbacks were both busy so nothing happens yet.
7. 5 seconds after 4: Callback 2 finishes sleeping and prints that it is done sleeping after receiving the first message
8. Immediately after 7: Callback 2 is given the second message now that it’s finished with the first and prints that it’s got that
9. 10 seconds after 3: Callback 1 finishes sleeping and prints that it is done sleeping after receiving the first message
10. Immediately after 9: Callback 1 is given the second message now that it’s finished with the first and prints that it’s got that
11. 5 seconds after 7: Callback 2 finishes sleeping and prints that it’s done sleeping after the second message

> 12. 10 seconds after 10: Callback 1 finishes sleeping and prints that it’s done sleeping after the second message

Reviewer: @jbarry-bdai meta: if you toss this into Gemini or GPT, you can get a nice [`mermaid` sequence diagram](https://mermaid.js.org/syntax/sequenceDiagram.html).
